qemu/tests/migration_multi_host_downtime_and_speed.py

```diff
@@ -65,57 +71,79 @@ def __init__(self, test, params, env):
-                s = vm.monitor.info("spice")
-                if isinstance(s, str):
-                    ret = "migrated: true" in s
-                else:
-                    ret = s.get("migrated") == "true"
-            o = vm.monitor.info("migrate")
-            if isinstance(o, str):
-                return ret and ("status: active" not in o)
-            else:
-                return ret and (o.get("status") != "active")
+        def clean_up(self, vm):
+            kill_bg_stress_cmd = params.get("kill_bg_stress_cmd",
+                                            "killall -9 stress")
+
+            logging.info("Kill the background stress test in the guest.")
+            session = vm.wait_for_login(timeout=self.login_timeout)
+            session.sendline(kill_bg_stress_cmd)
+            session.close()
+
+        @error.context_aware
+        def check_mig_downtime(self, vm):
+            logging.info("Check downtime after migration.")
+            actual_downtime = int(vm.monitor.info("migrate").get("downtime"))
+            if actual_downtime > self.mig_downtime * 1000:
+                error = "Migration failed for setting downtime, "
+                error += "Expected: '%d', Actual: '%d'" % (self.mig_downtime,
+                                                           actual_downtime)
+                raise error.TestFail(error)
+
+        @error.context_aware
+        def before_migration_downtime(self, mig_data):
+            if self.is_src:
+                vm = env.get_vm(params["main_vm"])
+                error.context("Set downtime before migration.", logging.info)
+                vm.monitor.migrate_set_downtime(self.mig_downtime)
+
+        @error.context_aware
+        def post_migration_before_downtime(self, vm, cancel_delay, mig_offline,
+                                           dsthost, vm_ports,
+                                           not_wait_for_migration,
+                                           fd, mig_data):
+            try:
+                vm.wait_for_migration(self.mig_timeout)
+            except virt_vm.VMMigrateTimeoutError:
+                raise error.TestFail("Migration failed with setting "
+                                     " downtime to %ds." % self.mig_downtime)
 
-        def wait_for_migration(self, vm, timeout):
-            if not utils_misc.wait_for(lambda: self.mig_finished(vm),
-                                       timeout,
-                                       2, 2,
-                                       "Waiting for migration to complete"):
-                raise virt_vm.VMMigrateTimeoutError("Timeout expired while"
-                                                    " waiting for migration"
-                                                    " to finish")
+            logging.info("Migration completed with downtime "
+                         "is %s seconds.", self.mig_downtime)
 
+            self.check_mig_downtime(vm)
+            vm.destroy(gracefully=False)
```

Comment:
for downtime cases, set not_wait_for_migration = yes, so need to destroy vm on src host.